src/storage/CompactionFilter.h

```diff
@@ -76,11 +79,57 @@ class StorageCompactionFilter final : public kvstore::KVFilter {
         return true;
     }
 
-    bool ttlValid(GraphSpaceID, const folly::StringPiece&) const {
-        // TODO(heng): Implement ttl filter after ttl schema pr merged in
+    bool ttlValid(GraphSpaceID spaceId, const folly::StringPiece& key,
+                  const folly::StringPiece& val) const {
+        if (NebulaKeyUtils::isVertex(key)) {
+            auto tagId = NebulaKeyUtils::getTagId(key);
+            auto schema = this->schemaMan_->getTagSchema(spaceId, tagId);
+            if (!schema) {
+                VLOG(3) << "Space " << spaceId << ", Tag " << tagId << " invalid";
+                return false;
+            }
+            auto reader = nebula::RowReader::getTagPropReader(schemaMan_, val, spaceId, tagId);
+            return checkDataTtlValid(schema.get(), reader.get());
+        } else if (NebulaKeyUtils::isEdge(key)) {
+            auto edgeType = NebulaKeyUtils::getEdgeType(key);
+            auto schema = this->schemaMan_->getEdgeSchema(spaceId, std::abs(edgeType));
+            if (!schema) {
+                VLOG(3) << "Space " << spaceId << ", EdgeType " << edgeType << " invalid";
+                return false;
+            }
+            auto reader = nebula::RowReader::getEdgePropReader(schemaMan_, val,
+                                                               spaceId, std::abs(edgeType));
+            return checkDataTtlValid(schema.get(), reader.get());
+        }
         return true;
     }
 
+    // TODO(panda) Optimize the method in the future
+    bool checkDataTtlValid(const meta::SchemaProviderIf* schema, nebula::RowReader* reader) const {
+        const meta::NebulaSchemaProvider* nschema =
+            dynamic_cast<const meta::NebulaSchemaProvider*>(schema);
+        if (nschema == NULL) {
+            return true;
+        }
```

Comment:
A question，Is the redundant check ？

Reply:
In some test case, it may be const meta::ResultSchemaProvider*